[Distributed] retain ad-hoc witnesses so that they are not optimized away in SIL

ktoso opened the pull request:
Since some of them (e.g. the decodeNextArgument specifically) are only used directly form IRGen, optimization thinks they are not used and may remove them. Making it impossible to use final classes (or structs sometimes?) to develop decoder types for distributed actors.

ktoso committed:
[Distributed] Retain adhoc decodeNextArgument in distributed thunk

ktoso commented:
resolves rdar://88211172 ([Distributed] SILGen must emit uses of decodeNextArgument so IRGen can get it cross module on `FINAL classes`)

ktoso commented on lib/Serialization/SILFormat.h:
```diff
@@ -282,6 +282,7 @@ namespace sil_block {
                      BCFixed<1>,  // is dynamically replacable
                      BCFixed<1>,  // exact self class
                      BCFixed<1>,  // is distributed
+//                     BCFixed<1>,  // is ad-hoc requirement witness
```
not sure if we need it here yet, probably we do

xedin commented on include/swift/SIL/SILFunction.h:
```diff
@@ -350,6 +362,9 @@ class SILFunction
   unsigned IsStaticallyLinked : 1;
 
+  /// The function is a witness to an ad-hoc requirement.
+  unsigned IsAdHocRequirementWitness : 1;
```
Do you even need this flag here, the check could be against pointer `AdHocRequirementFunction` which is _current_ not even set when `setDistributedAdHocRequirementFunction` is called...

xedin replied:
I think you need to do the same thing as `ReplacedFunction` where it's mangling is serialized into the attribute - https://github.com/apple/swift/blob/main/lib/SIL/Parser/ParseSIL.cpp#L1072-L1092

ktoso replied:
Hmm yeah very possible that I don't need it -- might have overdone the "same as `dynamic`", thanks I didn't spot that piece in ParseSIL for the replaced function!

xedin replied:
No worries!

xedin commented on lib/SIL/Parser/ParseSIL.cpp:
```diff
@@ -1006,6 +1007,8 @@ static bool parseDeclSILOptional(bool *isTransparent,
       *isDynamic = IsDynamic;
     else if (isDistributed && SP.P.Tok.getText() == "distributed")
       *isDistributed = IsDistributed;
+    else if (isDistributed && SP.P.Tok.getText() == "ad_hoc_requirement_witness")
```
Should check `isAdHocWitness` here

ktoso replied:
by the way I don't really understand where those tokens we match against comr from here hmmm

ktoso replied:
ok these come from SILPrinter

xedin commented on include/swift/SIL/SILFunction.h:
```diff
@@ -771,6 +812,13 @@ class SILFunction
   void
   setIsDistributed(IsDistributed_t value = IsDistributed_t::IsDistributed) {
+    if (value == IsDistributed_t::IsDistributed) {
+      assert(getName().equals("hello()"));
```
There is nothing obviously wrong in serialization that I can spot I guess the only way to check that would be to actually run it under lldb and check in what circumstances it gets set to true...

ktoso committed:
fix protocols

ktoso committed:
cleanups

ktoso commented:
@swift-ci please smoke test
@swift-ci please build toolchain macOS 

toolchain to verify in real project where this has shown up

ktoso commented on test/Distributed/Runtime/distributed_actor_cross_module_final_class_adhoc_requirement_not_optimized_away.swift:
```diff
@@ -0,0 +1,51 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift -target %target-cpu-apple-macosx10.15 -parse-as-library -emit-library -emit-module-path %t/FakeDistributedActorSystems.swiftmodule -module-name FakeDistributedActorSystems %S/../Inputs/FakeDistributedActorSystems.swift -o %t/%target-library-name(FakeDistributedActorSystems)
+// RUN: %target-build-swift -target %target-cpu-apple-macosx10.15 -parse-as-library -lFakeDistributedActorSystems -module-name main -I %t -L %t %s -o %t/a.out
+// RUN: %target-run %t/a.out | %FileCheck %s --color
```
It did reproduce after all I think

ktoso commented on include/swift/SIL/SILFunction.h:
```diff
@@ -212,6 +212,17 @@ class SILFunction
+  /// 'decodeNextArgument' which must be retained, as it is only used from IRGen
+  /// and such, appears as-if unused in SIL and would get optimized away.
+  // TODO: Consider making this a general "references adhoc functions" and make it an array?
+  SILFunction *RefAdHocRequirementFunction = nullptr;
```
for now just one attr; Want to make sure if works; if we really want to could become an array.

Do I need to make it array right away @xedin @aschwaighofer ?

ktoso commented:
@swift-ci please smoke test
@swift-ci please build toolchain macOS

toolchain to verify in real project where this has shown up

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please build toolchain macOS

aschwaighofer commented on lib/SIL/IR/SILFunctionBuilder.cpp:
```diff
@@ -200,29 +201,46 @@ void SILFunctionBuilder::addFunctionAttributes(
+              systemTy->getAnyNominal());
+      assert(decoderTy);
+
+      auto decodeFunc = C.getDecodeNextArgumentOnDistributedInvocationDecoder(
```
Nickpick. This is a lot of code inline. Does it make sense to sink the code into a method on actor (ClassDecl?)?

So that we have something like:
```
if (!constant.isDistributedThunk())
  return;
auto *actor = decl->getDeclContext()->getSelfNominalTypeDecl();
if (!actor)
  return;
auto decodeFunc = actor->ifActorGetDecodeNextArgumentOnDistributedInvocationDecode();
if (!decodeFunc)
  return;
auto decodeRef = SILDeclRef(decodeFunc);
...
```

ktoso replied:
Cool, yeah happy to do that -- I always (over)-worry about putting things on ClassDecl but that'll make it much nicer.
I'll refactor a bit, thanks for calling it out 👍

ktoso replied:
Long method name but nicer:

```
    if (auto decodeFunc =
            getAssociatedDistributedInvocationDecoderDecodeNextArgumentFunction(
                decl)) {

      auto decodeRef = SILDeclRef(decodeFunc);
      auto *adHocWitness = getOrCreateDeclaration(decodeFunc, decodeRef);
      F->setReferencedAdHocRequirementWitnessFunction(adHocWitness);
    }

```

aschwaighofer reviewed (approved):
lgtm

xedin commented on test/Distributed/Inputs/FakeDistributedActorSystems.swift:
```diff
@@ -300,7 +303,9 @@ public struct FakeInvocationEncoder : DistributedTargetInvocationEncoder {
-public class FakeInvocationDecoder : DistributedTargetInvocationDecoder {
+
+@available(SwiftStdlib 5.7, *)
+public final class FakeInvocationDecoder: DistributedTargetInvocationDecoder {
```
How hard would it be to add enum and struct variants for the decoder so all possibilities are tested? I think we could gyb this and some other tests too?

ktoso replied:
I'll make a radar for it but right now I think that's a bit of a pain... I have no idea about gyb tbh

xedin replied:
As soon as we are keeping track of that I am fine, I just don’t want it to be broken without us noticing…

ktoso replied:
Right yeah 👍

ktoso replied:
rdar://91735574

xedin reviewed (approved):
LGTM! I agree with @aschwaighofer that we should extract code for decode into a request but that could be done separately. More important I think to add tests for decoder being class/final class/struct/enum.

ktoso commented:
Cool, will add the tests. Manually is fine tbh...

ktoso committed:
review feedback, cleanup getting decode func

xedin commented on lib/SIL/IR/SILFunctionBuilder.cpp:
```diff
@@ -200,29 +201,35 @@ void SILFunctionBuilder::addFunctionAttributes(
+  } else if (constant.isDistributedThunk()) {
+    if (auto decodeFuncDecl =
+            getAssociatedDistributedInvocationDecoderDecodeNextArgumentFunction(
+                decl)) {
```
Nit: you can do that in the follow-up - if there is no `decodeNextArgument` we should fail here since the record of the thunk function is going to be ill-formed without it and that would lead to a crash in IRGen.

xedin commented on lib/SIL/IR/SILFunctionBuilder.cpp:
```diff
@@ -200,29 +201,35 @@ void SILFunctionBuilder::addFunctionAttributes(
+    if (auto decodeFuncDecl =
+            getAssociatedDistributedInvocationDecoderDecodeNextArgumentFunction(
+                decl)) {
+
```
Nit: extraneous newline.

xedin commented:
@swift-ci please test

xedin commented on test/SILGen/distributed_thunk.swift:
```diff
@@ -9,7 +9,7 @@ distributed actor DA {
 
 extension DA {
-  // CHECK-LABEL: sil hidden [thunk] [distributed] [ossa] @$s17distributed_thunk2DAC1fyyYaKFTE : $@convention(method) @async (@guaranteed DA) -> @error Error
+  // CHECK-LABEL: sil hidden [thunk] [distributed] [ref_adhoc_requirement_witness "$s11Distributed29LocalTestingInvocationDecoderC18decodeNextArgumentxyKSeRzSERzlF"] [ossa] @$s17distributed_thunk2DAC1fyyYaKFTE : $@convention(method) @async (@guaranteed DA) -> @error Error {
```
I think I can make a gyb to test that struct/class/final class/enum all have `ref_adhoc_requirement_witness` referenced after you land these changes.

ktoso replied:
Awesome, thanks -- I'll fix the SIL test now and land it then

ktoso committed:
[Distributed] Temporarily simplify SIL tests; avoid a few checks

ktoso commented:
@swift-ci please test

ktoso commented:
@swift-ci please test

ktoso commented:
This only fails in one of the Linux builds now... 

Could I ask you to help out here @xedin ? 🙏

xedin commented:
I should be able to take a look on Monday. Pretty sure that it’s not Linux specific, PR testing on macOS doesn’t run optimized tests if I remember correctly.

xedin committed:
[Distributed] SILOpt: Make sure that ad-hoc function does not get eliminated as unused

xedin committed:
[Distributed] IRGen: Remove a `decodeNextArgument` access workaround

The symbol is always retained now, so the workaround to use witness
table for non-final classes is no longer necessary.

xedin commented:
@swift-ci please test Linux platform

xedin committed:
[Distributed] NFC: Adjust accessor tests to check for direct call to `decodeNextArgument`

xedin commented:
@swift-ci please test

ktoso commented on lib/SILOptimizer/Analysis/CallerAnalysis.cpp:
```diff
@@ -33,8 +33,10 @@ CallerAnalysis::FunctionInfo::FunctionInfo(SILFunction *f)
-                             canBeCalledIndirectly(f->getRepresentation())),
+      mayHaveIndirectCallers(
+          f->getDynamicallyReplacedFunction() ||
+          f->getReferencedAdHocRequirementWitnessFunction() ||
```
Oh i see... makes sense

ktoso commented on lib/SILOptimizer/IPO/DeadFunctionElimination.cpp:
```diff
@@ -101,6 +101,9 @@ class DeadFunctionAndGlobalElimination {
       return true;
 
+    if (F->getReferencedAdHocRequirementWitnessFunction())
+      return true;
```
Ah okey so here too, I thought it was an either or but Makes sense!

ktoso commented on lib/SILOptimizer/IPO/DeadFunctionElimination.cpp:
```diff
@@ -325,6 +328,9 @@ class DeadFunctionAndGlobalElimination {
     LLVM_DEBUG(llvm::dbgs() << "    scan function " << F->getName() << '\n');
 
+    if (auto *adHocWitness = F->getReferencedAdHocRequirementWitnessFunction())
+      ensureAlive(adHocWitness);
```
<3

ktoso commented on lib/IRGen/GenDistributed.cpp:
```diff
@@ -813,22 +813,13 @@ ArgumentDecoderInfo DistributedAccessor::findArgumentDecoder(
+    /*secondaryValue=*/nullptr, signature);
 
   return {decoder, decoderTy, witnessTable, methodPtr, methodTy};
 }
```
Yess! 🥳

ktoso reviewed (commented):
Thank you so much @xedin ! 🙇
